File: src/EmployeeTasks/ListaCircular.java
package EmployeeTasks;
import java.util.Iterator;
import java.util.NoSuchElementException;

public class ListaCircular<T extends InterfazID> implements Iterable<T> {
    private Nodo<T> cabeza;
    private Nodo<T> cola;
    private int tamanio;

    public ListaCircular() {
        this.cabeza = null;
        this.cola = null;
        this.tamanio = 0;
    }

    public Nodo<T> getCabeza() {
        return cabeza;
    }

    public void agregar(T nuevo) {
        Nodo<T> nuevoNodo = new Nodo<>(nuevo);

        if (cabeza == null) {
            cabeza = nuevoNodo;
            cola = nuevoNodo;
            nuevoNodo.setSiguiente(cabeza);
        }

        else {
            cola.setSiguiente(nuevoNodo);
            nuevoNodo.setSiguiente(cabeza);
            cola = nuevoNodo;
        }
        tamanio++;
    }

    public T buscar(String id) {
        if (cabeza == null) {
            return null;
        }

        Nodo<T> actual = cabeza;

        do {
            if (actual.getElemento().getId().equals(id)) {
                return actual.getElemento();
            }
            actual = actual.getSiguiente();
        } while (actual != cabeza);

        return null;
    }

    public boolean eliminar(String id) {
        if (cabeza == null) {
            return false;
        }

        Nodo<T> actual = cabeza;
        Nodo<T> anterior = cola;

        do {
            if (actual.getElemento().getId().equals(id)) {
                if (actual == cabeza) {
                    cabeza = cabeza.getSiguiente();
                    cola.setSiguiente(cabeza);
                }

                else if (actual == cola) {
                    cola = anterior;
                    cola.setSiguiente(cabeza);
                }

                else {
                    anterior.setSiguiente(actual.getSiguiente());
                }
                tamanio--;
                return true;
            }
            anterior = actual;
            actual = actual.getSiguiente();
        } while (actual != cabeza);

        return false;
    }

    public int getTamanio() {
        return tamanio;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private Nodo<T> actual = cabeza;
            private Nodo<T> previo = null;
            private int elementosVisitados = 0;

            @Override
            public boolean hasNext() {
                return elementosVisitados < tamanio;
            }

            @Override
            public T next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                T elemento = actual.getElemento();
                previo = actual;
                actual = actual.getSiguiente();
                elementosVisitados++;
                return elemento;
            }

            @Override
            public void remove() {
                if (previo == null || elementosVisitados == 0) {
                    throw new IllegalStateException("Debe llamarse a next() antes de eliminar.");
                }

                ListaCircular.this.eliminar(previo.getElemento().getId());
                elementosVisitados--;
                previo = null;
            }
        };
    }
}

File: Modulo3_administrartrenes/ClasesControladoras3/ListaCircular.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProyectoEstructuras.Códigos.TareasEmpleado
{
    public class ListaCircular<T> : IEnumerable<T> where T : InterfazID
    {
        private Nodo<T> cabeza;
        private Nodo<T> cola;
        private int tamanio;

        public Nodo<T> Cabeza { get => cabeza; set => cabeza = value; }
        public Nodo<T> Cola { get => cola; set => cola = value; }
        public int Tamanio { get => tamanio; set => tamanio = value; }

        public ListaCircular()
        {
            this.cabeza = null;
            this.cola = null;
            this.tamanio = 0;
        }

        public void Agregar(T nuevo)
        {
            Nodo<T> nuevoNodo = new Nodo<T>(nuevo);

            if (cabeza == null)
            {
                cabeza = nuevoNodo;
                cola = nuevoNodo;
                nuevoNodo.Siguiente = cabeza;
            }
            else
            {
                cola.Siguiente = nuevoNodo;
                nuevoNodo.Siguiente = cabeza;
                cola = nuevoNodo;
            }

            tamanio++;
        }

        public T Buscar(string id)
        {
            if (cabeza == null)
            {
                return default;
            }

            Nodo<T> actual = cabeza;

            do
            {
                if (actual.Elemento.GetId().Trim().Equals(id.Trim(), StringComparison.OrdinalIgnoreCase))
                {
                    return actual.Elemento;
                }
                actual = actual.Siguiente;
            } while (actual != cabeza);

            return default;
        }

        public bool Eliminar(string id)
        {
            if (cabeza == null)
            {
                return false;
            }

            Nodo<T> actual = cabeza;
            Nodo<T> anterior = cola;

            do
            {
                if (actual.Elemento.GetId().Equals(id))
                {
                    if (actual == cabeza)
                    {
                        cabeza = cabeza.Siguiente;
                        cola.Siguiente = cabeza;
                    }
                    else if (actual == cola)
                    {
                        cola = anterior;
                        cola.Siguiente = cabeza;
                    }
                    else
                    {
                        anterior.Siguiente = actual.Siguiente;
                    }

                    tamanio--;
                    return true;
                }

                anterior = actual;
                actual = actual.Siguiente;
            } while (actual != cabeza);

            return false;
        }

        public bool Modificar(string id, T nuevoElemento)
        {
            if (cabeza == null)
            {
                return false;
            }

            Nodo<T> actual = cabeza;

            do
            {
                if (actual.Elemento.GetId().Equals(id))
                {
                    actual.Elemento = nuevoElemento;
                    return true;
                }
                actual = actual.Siguiente;
            } while (actual != cabeza);

            return false;
        }

        public IEnumerator<T> GetEnumerator()
        {
            if (cabeza == null)
            {
                yield break; // Devolver un iterador vacío si la lista está vacía
            }

            Nodo<T> actual = cabeza;
            int elementosCedidos = 0; // Contador para saber cuántos elementos hemos "devuelto"

            // El bucle debe ejecutarse exactamente Tamanio veces para visitar cada nodo una vez
            while (elementosCedidos < tamanio)
            {
                yield return actual.Elemento;
                actual = actual.Siguiente;
                elementosCedidos++;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }
}
